File: generated_code/python/solution_297_3.py
from collections import deque

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:
    """
    Algorithm: Preorder Traversal (DFS)
    
    Time Complexity: 
      - serialize: O(N) where N is the number of nodes, as we visit each node once.
      - deserialize: O(N) since we process each split token exactly once.
    Space Complexity:
      - serialize: O(N) to store the call stack and the resulting string pieces.
      - deserialize: O(N) to store the list of tokens and the recursion stack.
    """

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        vals = []
        
        def dfs(node):
            if not node:
                vals.append("#")
                return
            vals.append(str(node.val))
            dfs(node.left)
            dfs(node.right)
            
        dfs(root)
        return ",".join(vals)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        # Split the serialized string into an iterable queue of tokens
        tokens = deque(data.split(","))
        
        def dfs():
            if not tokens:
                return None
            
            val = tokens.popleft()
            if val == "#":
                return None
            
            # Reconstruct the current node and its subtrees sequentially
            node = TreeNode(int(val))
            node.left = dfs()
            node.right = dfs()
            return node
            
        return dfs()

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: generated_code/cpp/solution_297_3.cpp
#include <string>
#include <sstream>
#include <vector>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Codec {
private:
    // Helper function for serialization using pre-order traversal (DFS).
    void serializeDFS(TreeNode* root, std::ostringstream& out) {
        if (!root) {
            out << "# "; // Use '#' to represent a null node, separated by a space
            return;
        }
        out << root->val << " ";
        serializeDFS(root->left, out);
        serializeDFS(root->right, out);
    }

    // Helper function for deserialization using pre-order traversal (DFS).
    TreeNode* deserializeDFS(std::istringstream& in) {
        std::string val_str;
        if (!(in >> val_str)) {
            return nullptr;
        }
        
        if (val_str == "#") {
            return nullptr;
        }

        // Reconstruct the current node and its subtrees recursively
        TreeNode* root = new TreeNode(std::stoi(val_str));
        root->left = deserializeDFS(in);
        root->right = deserializeDFS(in);
        return root;
    }

public:
    // Encodes a tree to a single string.
    // Time Complexity: O(N) where N is the number of nodes.
    // Space Complexity: O(N) for the recursion stack and output stream.
    std::string serialize(TreeNode* root) {
        std::ostringstream out;
        serializeDFS(root, out);
        return out.str();
    }

    // Decodes your encoded data to tree.
    // Time Complexity: O(N) where N is the number of nodes.
    // Space Complexity: O(N) for the recursion stack and input stream.
    TreeNode* deserialize(std::string data) {
        std::istringstream in(data);
        return deserializeDFS(in);
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));